libs/wingcompiler/src/compile.ts

```diff
@@ -259,57 +250,116 @@ function isEntrypointFile(path: string) {
+    } catch (error) {
+      // This is a bug in the compiler, indicate a compilation failure.
+      // The bug details should be part of the diagnostics handling below.
+      compileSuccess = false;
+    }
+    if (!compileSuccess) {
+      // This is a bug in the user's code. Print the compiler diagnostics.
+      throw new CompileError(errors);
+    }
+
+    return join(props.workDir, WINGC_PREFLIGHT);
+  }
+}
+
 async function runPreflightCodeInVm(
-  workDir: string,
+  entrypoint: string,
   wingDir: string,
-  tempProcess: { env: Record<string, string | undefined> },
+  env: Record<string, string | undefined>,
   log?: (...args: any[]) => void
 ): Promise<void> {
-  const artifactPath = join(workDir, WINGC_PREFLIGHT);
-  log?.("reading artifact from %s", artifactPath);
-  const artifact = await fs.readFile(artifactPath, "utf-8");
+  log?.("reading artifact from %s", entrypoint);
+  const artifact = await fs.readFile(entrypoint, "utf-8");
   log?.("artifact: %s", artifact);
 
   // Try looking for dependencies not only in the current directory (wherever
   // the wing CLI was installed to), but also in the source code directory.
   // This is necessary because the Wing app may have installed dependencies in
   // the project directory.
   const requireResolve = (path: string) =>
-    require.resolve(path, { paths: [__dirname, wingDir, workDir] });
+    require.resolve(path, { paths: [__dirname, wingDir, dirname(entrypoint)] });
   const preflightRequire = (path: string) => require(requireResolve(path));
   preflightRequire.resolve = requireResolve;
 
-  // If you're wondering how the execution of the preflight works, despite it
-  // being in a different directory: it works because at the top of the file
-  // require.resolve is called to cache wingsdk in-memory. So by the time VM
-  // is starting up, the passed context already has wingsdk in it.
-  // "__dirname" is also synthetically changed so nested requires work.
-  const context = vm.createContext({
-    require: preflightRequire,
-    process: tempProcess,
-    console,
-    __dirname: workDir,
-    __filename: artifactPath,
-    // since the SDK is loaded in the outer VM, we need these to be the same class instance,
-    // otherwise "instanceof" won't work between preflight code and the SDK. this is needed e.g. in
-    // `serializeImmutableData` which has special cases for serializing these types.
-    Map,
-    Set,
-    Array,
-    Promise,
-    Object,
-    RegExp,
-    String,
-    Date,
-    Function,
-  });
-
   try {
-    vm.runInContext(artifact, context, {
-      filename: artifactPath
+    await new Promise((resolve, reject) => {
+      const worker = new Worker(entrypoint, {
+        env,
+      });
+      worker.on("error", reject);
+      worker.on("exit", (code) => {
+        if (code === 0) {
+          resolve(undefined);
+        } else {
+          reject(new Error(`Worker stopped with exit code ${code}`));
+        }
+      });
```

Comment:
Where do worker error logs go? I guess they inherit the stderr of the process it's created in, so if preflight code crashes the stack trace should still get printed to the CLI like we have today?

Reply:
If there's an uncaught exception, the "error" event gets raised and rejects from this promise.
stderr is inherited, so that works the same as when we used a vm.

The non-zero exit code here would only happen if someone explicitly called `process.exit(non-zero)` or something similar.